### PR TITLE
PWGHF: implementation of LcToK0Sp ML responce class and corresponding changes in the LcToK0Sp canidate selector

### PR DESCRIPTION
**Pull request by emeninno**

### CODE REVIEW

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+// possible input features for ML
+enum class InputFeaturesLcToK0sP : uint8_t {
+  posX,
+  posY,
+  posZ,
+  xSecondaryVertex,
+  ySecondaryVertex,
+  zSecondaryVertex,
+  errorDecayLength,
+  errorDecayLengthXY,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:72]: These variables are typically not used as ML features, so I would remove them
```suggestion
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  impactParameterNormalised0,
+  ptProng0,
+  impactParameterNormalised1,
+  ptProng1,
+  pxProng0,
+  pyProng0,
+  pzProng0,
+  pxProng1,
+  pyProng1,
+  pzProng1,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:88]: No need to have momentum components
```suggestion
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  impactParameter0,
+  impactParameter1,
+  errorImpactParameter0,
+  errorImpactParameter1,
+  v0X,
+  v0Y,
+  v0Z,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:95]: No need to have these as well
```suggestion
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  v0cosPA,
+  v0MLambda,
+  v0MAntiLambda,
+  v0MK0Short,
+  v0MGamma,
+  ctV0,
+  dcaV0daughters,
+  pxPos,
+  pyPos,
+  pzPos,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:106]: No need to have momentum components
```suggestion
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  ptV0Pos,
+  dcaPosToPV,
+  pxNeg,
+  pyNeg,
+  pzNeg,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:111]: No need to have momentum components
```suggestion
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+  ptV0Neg,
+  dcaNegToPV,
+  nSigmaTPCPr0,
+  nSigmaTOFPr0,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:115]: I would also add the combined nSigma variable (see e.g. https://github.com/AliceO2Group/O2Physics/blob/524816143c1d9d4b3ce6f85d6c2e7eed7372bc25/PWGHF/Core/HfMlResponseD0ToKPi.h#L98-L99) Also, don't forget to spawn the tables of the combined nSigma tables (see https://github.com/AliceO2Group/O2Physics/pull/3872)

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+  m,
+  pt,
+  p,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:118]: Mass and momentum of the Lc should never been used in a ML training
```suggestion
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+  cpa,
+  cpaXY,
+  ct,
+  eta,
+  phi,
+  y,
+  e
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:125]: These variables should not be used as training variables
```suggestion
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+  /// \return inputFeatures vector
+  template <typename T1, typename T2>
+  std::vector<float> getInputFeatures(T1 const& candidate,
+                                      T2 const& bach)
+  {
+    std::vector<float> inputFeatures;
+
+    for (const auto& idx : MlResponse<TypeOutputScore>::mCachedIndices) {
+      switch (idx) {
+        CHECK_AND_FILL_VEC_LC(posX);
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:151]: Remove from here the variables that I suggested to remove from the `enum`

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    return inputFeatures;
+  }
+
+ protected:
+  /// Method to fill the map of available input features
+  void setAvailableInputFeatures()
+  {
+    MlResponse<TypeOutputScore>::mAvailableInputFeatures = {
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:225]: Remove from here the variables that I suggested to remove from the `enum`

**Review by fgrosa — CHANGES_REQUESTED** [review]
Hi @emeninno! Thanks for the development, I have comments mostly on the ML features set, since several of them should not be used in the training so it would be better to remove them to avoid that less expert analysers use them.

**Comment by fgrosa** [comment]
@emeninno when you implement the comments rebase on master because of [this PR](https://github.com/AliceO2Group/O2Physics/pull/3952), which fixed a bug in the retrieving of BDT models from CCDB.

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+  v0MGamma,
+  ctV0,
+  dcaV0daughters,
+  ptV0Pos,
+  dcaPosToPV,
+  ptV0Neg,
+  dcaNegToPV,
+  nSigmaTPCPr0,
+  nSigmaTOFPr0,
+  nSigmaTpcTOFPr0,
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:91]:
```suggestion
  nSigmaTpcPr0,
  nSigmaTofPr0,
  nSigmaTpcTofPr0,
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+        CHECK_AND_FILL_VEC_LC_FULL(candidate, dcaNegToPV, dcanegtopv);
+        CHECK_AND_FILL_VEC_LC(cpa);
+        CHECK_AND_FILL_VEC_LC(cpaXY);
+        CHECK_AND_FILL_VEC_LC_HFHELPER(candidate, ct, ctLc);
+        // TPC PID variables
+        CHECK_AND_FILL_VEC_LC_FULL(bach, nSigmaTPCPr0, tpcNSigmaPr);
+        // TOF PID variables
+        CHECK_AND_FILL_VEC_LC_FULL(bach, nSigmaTOFPr0, tofNSigmaPr);
+        // Combined nSigma variable
+        CHECK_AND_FILL_VEC_LC_FULL(bach, nSigmaTpcTOFPr0, tpcTofNSigmaPr);
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:153]:
```suggestion
        CHECK_AND_FILL_VEC_LC_FULL(bach, nSigmaTpcPr0, tpcNSigmaPr);
        // TOF PID variables
        CHECK_AND_FILL_VEC_LC_FULL(bach, nSigmaTofPr0, tofNSigmaPr);
        // Combined nSigma variable
        CHECK_AND_FILL_VEC_LC_FULL(bach, nSigmaTpcTofPr0, tpcTofNSigmaPr);
```

File: PWGHF/Core/HfMlResponseLcToK0sP.h [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+      FILL_MAP_LC(dcaNegToPV),
+      FILL_MAP_LC(cpa),
+      FILL_MAP_LC(cpaXY),
+      FILL_MAP_LC(ct),
+      // TPC PID variables
+      FILL_MAP_LC(nSigmaTPCPr0),
+      // TOF PID variables
+      FILL_MAP_LC(nSigmaTOFPr0),
+      // Combined nSigma variable
+      FILL_MAP_LC(nSigmaTpcTOFPr0)};
```
Comment (fgrosa) [inline PWGHF/Core/HfMlResponseLcToK0sP.h:197]:
```suggestion
      FILL_MAP_LC(nSigmaTpcPr0),
      // TOF PID variables
      FILL_MAP_LC(nSigmaTofPr0),
      // Combined nSigma variable
      FILL_MAP_LC(nSigmaTpcTofPr0)};
```

**Review by fgrosa — CHANGES_REQUESTED** [review]
thanks for implementing my suggestions, I just have few minor comments on the variable names (you can just click on "commit suggestion"), then we can merge it!

**Review by fgrosa — APPROVED** [review]

**Comment by emeninno** [comment]
Hi @fgrosa , thanks a lot for your suggestions. I implemented them.
Cheers
Elisa

**Pull request merged**